undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterAdd.java

```diff
@@ -54,12 +53,20 @@ class FilterAdd extends AbstractAddStepHandler {
     @Override
     protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
         final String name = context.getCurrentAddressValue();
-        final ServiceTarget target = context.getServiceTarget();
-        final ServiceBuilder<?> sb = target.addService(UndertowService.FILTER.append(name));
-        final Consumer<HandlerWrapper> serviceConsumer = sb.provides(UndertowService.FILTER.append(name));
+        CapabilityServiceBuilder<?> csb = context.getCapabilityServiceTarget().addCapability(FilterCapabilities.FILTER_CAPABILITY.getDefinition());
+        Consumer<HandlerWrapper> serviceConsumer =  csb.provides(FilterCapabilities.FILTER_CAPABILITY.getDefinition(), UndertowService.FILTER.append(name));
         HandlerWrapper wrapper = this.factory.createHandlerWrapper(context, model);
-        sb.setInstance(Service.newInstance(serviceConsumer, wrapper));
-        sb.setInitialMode(ServiceController.Mode.ON_DEMAND);
-        sb.install();
+        csb.setInstance(new FilterProviderService(serviceConsumer, wrapper));
```

Comment:
Why introduce a redundant FilterProviderService class?  The implementation is identical to the previous logic.